cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -293,76 +315,172 @@ func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
 }
 
 type PaginationOptions struct {
-	Cursor *events.Cursor `json:"cursor,omitempty"`
-	Limit  uint           `json:"limit,omitempty"`
+	Cursor *db.Cursor `json:"cursor,omitempty"`
+	Limit  uint       `json:"limit,omitempty"`
 }
 
 type GetEventsResponse struct {
 	Events       []EventInfo `json:"events"`
 	LatestLedger uint32      `json:"latestLedger"`
 }
 
-type eventScanner interface {
-	Scan(eventRange events.Range, f events.ScanFunction) (uint32, error)
-}
-
 type eventsRPCHandler struct {
-	scanner      eventScanner
+	dbReader     db.EventReader
 	maxLimit     uint
 	defaultLimit uint
+	logger       *log.Entry
+	ledgerReader db.LedgerReader
 }
 
-func (h eventsRPCHandler) getEvents(request GetEventsRequest) (GetEventsResponse, error) {
+func combineContractIDs(filters []EventFilter) ([][]byte, error) {
+	contractIDSet := set.NewSet[string](maxFiltersLimit * maxContractIDsLimit)
+
+	for _, filter := range filters {
+		for _, contractID := range filter.ContractIDs {
+			contractIDSet.Add(contractID)
+		}
+	}
+
+	contractIDs := make([][]byte, 0, len(contractIDSet))
+	for contractID := range contractIDSet {
+		id, err := strkey.Decode(strkey.VersionByteContract, contractID)
+		if err != nil {
+			return nil, fmt.Errorf("invalid contract ID %v: ", contractID)
+		}
+
+		contractIDs = append(contractIDs, id)
+	}
+	return contractIDs, nil
+}
+
+func combineEventTypes(filters []EventFilter) []int {
+	eventTypes := make(map[int]bool)
+	for _, filter := range filters {
+		for _, eventType := range filter.EventType.Keys() {
+			eventTypeXDR := getEventTypeXDRFromEventType()[eventType]
+			eventTypes[int(eventTypeXDR)] = true
+		}
+	}
+	uniqueEventTypes := make([]int, 0, len(eventTypes))
+	for eventType := range eventTypes {
+		uniqueEventTypes = append(uniqueEventTypes, eventType)
+	}
+	return uniqueEventTypes
+}
+
+func combineTopics(filters []EventFilter) ([][][]byte, error) {
+	encodedTopicsList := make([][][]byte, db.MaxTopicCount)
+
+	for _, filter := range filters {
+		if len(filter.Topics) == 0 {
+			return [][][]byte{}, nil
+		}
+
+		for _, topicFilter := range filter.Topics {
+			for i, segmentFilter := range topicFilter {
+				if segmentFilter.wildcard == nil && segmentFilter.scval != nil {
+					encodedTopic, err := segmentFilter.scval.MarshalBinary()
+					if err != nil {
+						return [][][]byte{}, fmt.Errorf("failed to marshal segment: %w", err)
+					}
+					encodedTopicsList[i] = append(encodedTopicsList[i], encodedTopic)
```

Comment:
Wait how is this an array of an array of an array of bytes? Since you are going to pass all of them to the SQL flattened out, shouldn't it only be a 2D array? As in, a 4-element array of topics `{ {topic1, topic1, ...}, {topic2, ... } }`?

Reply:
` { {topic1, topic1, ...}, {topic2, ... } }` --> Array of Array of String --> Array of Array of Array of byte

Reply:
![image](https://github.com/user-attachments/assets/81d14687-37e2-4771-a4d9-029208578263)

Reply:
Oh duh :facepalm: thanks lol